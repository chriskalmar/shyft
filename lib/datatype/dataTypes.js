
import DataType, {
  DataTypeUser,
} from './DataType';



export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user'
})



export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs'
})


export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values'
})


export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values'
})


export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values'
})


export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values'
})


export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values'
})


export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects'
})


export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp'
})


export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information'
})
