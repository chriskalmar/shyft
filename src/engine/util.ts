import * as casual from 'casual';
import * as _ from 'lodash';
import { viewAttributePropertiesWhitelist } from './constants';

type StringFunction = () => string;

export const deleteUndefinedProps = (obj) =>
  Object.keys(obj).forEach((key) => obj[key] === undefined && delete obj[key]);

export const passOrThrow = (
  condition: any,
  messageFn: string | StringFunction,
): void => {
  // providing the error message as a callback return massively improves code coverage reports.
  // the message function is only evaluated if the condition fails, which will show up correctly
  // in the coverage reports.
  if (typeof messageFn !== 'function') {
    throw new Error('passOrThrow() expects messageFn to be a function');
  }
  if (!condition) {
    throw new Error(messageFn());
  }
};

export const resolveFunctionMap = (functionOrMap) => {
  return typeof functionOrMap === 'function'
    ? functionOrMap()
    : { ...functionOrMap };
};

export const isMap = (map: object, nonEmpty?: boolean): boolean => {
  return (
    map !== null &&
    typeof map === 'object' &&
    Array.isArray(map) === false &&
    (!nonEmpty || (nonEmpty && Object.keys(map).length > 0))
  );
};

export const isFunction = (fn: any): boolean => {
  return typeof fn === 'function';
};

export const isArray = (set: any[], nonEmpty?: boolean): boolean => {
  return (
    set !== null &&
    Array.isArray(set) === true &&
    (!nonEmpty || (nonEmpty && set.length > 0))
  );
};

export const isString = (str: any): boolean => typeof str === 'string';

export const mergeMaps = (first, second) => {
  if (!isMap(first) || !isMap(second)) {
    throw new Error('mergeMaps() expects 2 maps for a merge to work');
  }

  return { ...first, ...second };
};

export const mapOverProperties = (
  object: object,
  iteratee: (val: any, key: any) => any,
) => {
  passOrThrow(isMap(object), () => 'Provided object is not a map');

  passOrThrow(
    isFunction(iteratee),
    () => 'Provided iteratee is not a function',
  );

  const keys = Object.keys(object);

  keys.forEach((key) => {
    iteratee(object[key], key);
  });
};

export const sortDataByKeys = (
  keys: string[],
  data: object[],
  keyProperty = 'id',
) => {
  const map = {};
  const result = [];
  const order = {};

  if (!keys || (isArray(keys) && keys.length === 0)) {
    return [];
  }

  if (!data || (isArray(data) && data.length === 0)) {
    return keys.map(() => null);
  }

  for (let d = 0; d < data.length; d++) {
    const row = data[d];
    const id = row[keyProperty];
    const found = map[id];

    if (!found) {
      map[id] = row;
    } else if (found instanceof Array) {
      found.push(row);
    } else {
      map[id] = [found, row];
    }
  }

  for (let k = 0; k < keys.length; k++) {
    const key = keys[k];
    const found = map[key];

    if (typeof found === 'undefined') {
      result.push(null);
    } else if (found instanceof Array) {
      let idx = order[key];

      if (typeof idx === 'undefined') {
        idx = 0;
      } else {
        idx++;
      }

      order[key] = idx;

      result.push(found[idx]);
    } else {
      result.push(found);
    }
  }

  return result;
};

export const reverseString = (str) => {
  return str.split('').reverse().join('');
};

export const randomJson = () => {
  const ret = {};
  const count = casual.integer(1, 5);

  _.times(count, () => {
    const key = _.camelCase(casual.words(2));
    const value = Math.random() > 0.5 ? casual.title : casual.integer();

    ret[key] = value;
  });

  return ret;
};

export const asyncForEach = async (
  array: any[],
  callback: (val, idx, arr) => any,
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const isDefined = (val: any): boolean => typeof val !== 'undefined';

export const convertEntityToViewAttribute = (attribute) => {
  if (!_.isObject(attribute)) {
    throw new Error(
      'convertEntityToViewAttribute() expects an attribute as input',
    );
  }

  const newAttribute = {};

  viewAttributePropertiesWhitelist.map((prop) => {
    newAttribute[prop] = attribute[prop];
  });

  return newAttribute;
};

export const convertEntityToViewAttributesMap = (attributesMap) => {
  const newAttributesMap = {};

  if (!_.isObject(attributesMap)) {
    throw new Error(
      'convertEntityToViewAttributesMap() expects an attributes map as input',
    );
  }

  for (const [name, attribute] of Object.entries(attributesMap)) {
    newAttributesMap[name] = convertEntityToViewAttribute(attribute);
  }

  return newAttributesMap;
};
