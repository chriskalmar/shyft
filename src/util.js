import crypto from 'crypto';

// postgres limitation
const SYSTEM_NAME_MAX_LENGTH = 63;

export const generateIndexName = (entityName, attributes, suffix = '_idx') => {
  let ret = `${entityName}_${suffix}_`;
  const attributesChain = (attributes || []).join('_');
  const attributesSortedChain = (attributes || []).sort().join('_');

  if (ret.length + attributesChain.length > SYSTEM_NAME_MAX_LENGTH) {
    const hashLength = 10;
    const hash = crypto
      .createHash('sha256')
      .update(attributesSortedChain)
      .digest('hex')
      .substr(0, hashLength);

    const shortendAttributesChain = attributesChain.substr(
      0,
      SYSTEM_NAME_MAX_LENGTH - hashLength - ret.length - 1,
    );

    ret = `${entityName}_${shortendAttributesChain}_${hash}_${suffix}`;
  } else {
    ret = `${entityName}_${attributesChain}_${suffix}`;
  }

  return ret;
};

export const invertDirection = direction => {
  return direction === 'DESC' ? 'ASC' : 'DESC';
};

export const getLimit = (first, last) => {
  if (first >= 0) {
    return first;
  } else if (last >= 0) {
    return last;
  }

  return 10;
};

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array);
  }
};

export const quote = item => {
  if (!item) {
    throw new Error('quote() requires an input');
  }

  // special care for json pointers
  if (item.includes('->')) {
    const arr = item.split('->');
    arr[0] = `"${arr[0]}"`;
    return arr.join('->');
  }

  return `"${item}"`;
};
