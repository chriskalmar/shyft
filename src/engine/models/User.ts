import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { Index, INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';

export const User = new Entity({
  name: 'User',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],

  attributes: {
    username: {
      type: DataTypeString,
      description: 'Username',
      required: true,
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
    },

    language: {
      type: Language,
      description: 'User language',
      required: true,
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true,
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true,
    },
  },
});
