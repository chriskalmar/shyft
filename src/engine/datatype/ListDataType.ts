import { passOrThrow, isFunction, isArray } from '../util';
import { Entity, isEntity } from '../entity/Entity';
import { DataType, isDataType, DataTypeFunction } from './DataType';
import { ComplexDataType, isComplexDataType } from './ComplexDataType';

export type ListDataTypeSetupType = {
  name: string;
  description: string;
  itemType: Entity | ComplexDataType | DataTypeFunction;
  minItems?: number;
  maxItems?: number;
};

export class ListDataType extends ComplexDataType {
  name: string;
  description: string;
  itemType: Entity | ComplexDataType | DataTypeFunction;
  _itemType: DataType | ComplexDataType;
  minItems?: number;
  maxItems?: number;

  constructor(setup: ListDataTypeSetupType = {} as ListDataTypeSetupType) {
    super();

    const { name, description, itemType, minItems, maxItems } = setup;

    passOrThrow(name, () => 'Missing list data type name');
    passOrThrow(
      description,
      () => `Missing description for list data type '${name}'`,
    );
    passOrThrow(
      itemType,
      () => `Missing item type for list data type '${name}'`,
    );

    passOrThrow(
      isDataType(itemType) ||
        isEntity(itemType) ||
        isComplexDataType(itemType) ||
        isFunction(itemType),
      () =>
        `List data type '${name}' has invalid item type '${String(itemType)}'`,
    );

    const _minItems = minItems || 0;
    const _maxItems = maxItems || 0;

    if (_minItems) {
      passOrThrow(
        Number.isInteger(_minItems) && _minItems >= 0,
        () =>
          `List data type '${name}' has invalid minItems setting '${_minItems}'`,
      );
    }

    if (_maxItems) {
      passOrThrow(
        Number.isInteger(_maxItems) && _maxItems >= 0,
        () =>
          `List data type '${name}' has invalid maxItems setting '${_maxItems}'`,
      );
    }

    passOrThrow(
      _minItems <= _maxItems || _maxItems === 0,
      () =>
        `List data type '${name}' has a bigger minItems than the maxItems setting`,
    );

    this.name = name;
    this.description = description;
    this.itemType = itemType;
    this.minItems = _minItems;
    this.maxItems = _maxItems;
  }

  _processItemType(): DataType | ComplexDataType {
    if (isFunction(this.itemType)) {
      const itemTypeBuilder: DataTypeFunction = this
        .itemType as DataTypeFunction;
      const itemType = itemTypeBuilder({
        name: this.name,
        description: this.description,
      });

      passOrThrow(
        isDataType(itemType) ||
          isEntity(itemType) ||
          isComplexDataType(itemType),
        () =>
          `List data type '${
            this.name
          }' has invalid dynamic item type '${String(itemType)}'`,
      );

      return itemType;
    }

    return this.itemType;
  }

  getItemType(): DataType | ComplexDataType {
    if (this._itemType) {
      return this._itemType;
    }

    const ret = (this._itemType = this._processItemType());
    return ret;
  }

  validate = (payload: any): void => {
    if (payload) {
      passOrThrow(
        isArray(payload),
        () => `List data type '${this.name}' expects an array of items`,
      );

      passOrThrow(
        payload.length >= this.minItems,
        () =>
          `List data type '${this.name}' requires a minimum of ${this.minItems} items`,
      );

      passOrThrow(
        this.maxItems === 0 || payload.length <= this.maxItems,
        () =>
          `List data type '${this.name}' requires a maximum of ${this.maxItems} items`,
      );
    }
  };

  toString(): string {
    return this.name;
  }
}

export const isListDataType = (obj: any): boolean => {
  return obj instanceof ListDataType;
};

export const buildListDataType = (obj: {
  itemType: any;
  // itemType: Entity | ComplexDataType | DataTypeFunction;
}): Function => {
  return ({ name, description }): ListDataType =>
    new ListDataType({
      description,
      ...obj,
      name,
    });
};
