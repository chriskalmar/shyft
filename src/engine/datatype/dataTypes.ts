import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
