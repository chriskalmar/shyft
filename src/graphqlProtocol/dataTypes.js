import { GraphQLScalarType, Kind, GraphQLError } from 'graphql';
import { serializeCursor, deserializeCursor } from './util';
import * as dateFns from 'date-fns';

const dateTimeRegex = /^\d+-[0-9]{2}-[0-9]{2}[\sT]([0-1][0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?(\s?[+-]([0-1][0-9]|2[0-3])(:[0-5][0-9])?)?$/;
const timeRegex = /^([0-1][0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?(\s?[+-]([0-1][0-9]|2[0-3])(:[0-5][0-9])?)?$/;

export { GraphQLJSON } from 'graphql-type-json';

export const GraphQLCursor = new GraphQLScalarType({
  name: 'Cursor',
  description:
    'A cursor to define a location in a data set used for pagination.',
  serialize: serializeCursor,
  parseValue: deserializeCursor,
  parseLiteral(ast) {
    return ast.kind === Kind.STRING ? deserializeCursor(ast.value) : null;
  },
});

export const GraphQLBigInt = new GraphQLScalarType({
  name: 'BigInt',
  description:
    'The `BigInt` scalar type represents a 64-bit integer. As JavaScript ' +
    'is limited to a precision of 53 bits on integer values, all `BigInt` ' +
    'values will be output as strings.',
  serialize: String,
  parseValue: String,
  parseLiteral(ast) {
    return ast.kind === Kind.STRING ? ast.value : null;
  },
});

export const GraphQLDateTime = new GraphQLScalarType({
  name: 'DateTime',
  description: 'The `DateTime` scalar type represents a date and time string.',
  serialize: value => {
    return value;
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Query error: Can only parse strings but got a: ${ast.kind}`,
        [ ast ],
      );
    }

    if (!dateTimeRegex.test(ast.value)) {
      throw new GraphQLError(
        'Query error: String is not a valid date time string',
        [ ast ],
      );
    }

    const dateString = ast.value.substring(0, 10);
    const asDate = dateFns.parse(dateString);

    if (
      !dateFns.isValid(asDate) ||
      dateString !== dateFns.format(asDate, 'YYYY-MM-DD')
    ) {
      throw new GraphQLError(
        'Query error: String is not a valid date time string',
        [ ast ],
      );
    }

    return ast.value;
  },
});

export const GraphQLDate = new GraphQLScalarType({
  name: 'Date',
  description: 'The `Date` scalar type represents a date string.',
  serialize: value => {
    return value;
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Query error: Can only parse strings but got a: ${ast.kind}`,
        [ ast ],
      );
    }

    const asDate = dateFns.parse(ast.value);

    if (
      !dateFns.isValid(asDate) ||
      ast.value !== dateFns.format(asDate, 'YYYY-MM-DD')
    ) {
      throw new GraphQLError('Query error: String is not a valid date string', [
        ast,
      ]);
    }

    return ast.value;
  },
});

export const GraphQLTime = new GraphQLScalarType({
  name: 'Time',
  description: 'The `Time` scalar type represents a time string.',
  serialize: value => {
    return value;
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Query error: Can only parse strings but got a: ${ast.kind}`,
        [ ast ],
      );
    }

    if (!timeRegex.test(ast.value)) {
      throw new GraphQLError('Query error: String is not a valid time string', [
        ast,
      ]);
    }

    return ast.value;
  },
});
