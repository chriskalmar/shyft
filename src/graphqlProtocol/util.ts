import * as _ from 'lodash';
import * as pluralize from 'pluralize';
import { RELAY_TYPE_PROMOTER_FIELD } from './protocolGraphqlConstants';
import { CustomError } from '../engine/CustomError';
import { isDefined } from '../engine/util';
import { getRegisteredEntityAttribute } from './registry';

export function generateTypeName(name) {
  return _.camelCase(name);
}

export function pascalCase(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function plural(str) {
  return pluralize.plural(str);
}

export function generateTypeNamePascalCase(name) {
  return pascalCase(generateTypeName(name));
}

export function generateTypeNamePlural(name) {
  return pluralize.plural(generateTypeName(name));
}

export function generateTypeNamePluralPascalCase(name) {
  return pascalCase(pluralize.plural(generateTypeName(name)));
}

export function generateTypeNameUpperCase(name) {
  return _.snakeCase(name).toUpperCase();
}

export const toBase64 = (value) =>
  new Buffer(value.toString()).toString('base64');
export const fromBase64 = (value) =>
  new Buffer(value.toString(), 'base64').toString();

export const serializeCursor = (cursor) => toBase64(JSON.stringify(cursor));
export const deserializeCursor = (cursor) => {
  try {
    return JSON.parse(fromBase64(cursor));
  } catch (e) {
    // TODO: check
    // should throw error, not return
    // https://github.com/graphql/graphql-js/issues/910
    return new Error('Invalid cursor provided');
  }
};

export const addRelayTypePromoterToInstance = (typeName, instance) => {
  if (!instance) {
    return instance;
  }

  instance[RELAY_TYPE_PROMOTER_FIELD] = typeName;
  return instance;
};

export const addRelayTypePromoterToList = (typeName, list) => {
  return list.map((instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  });
};

export const addRelayTypePromoterToInstanceFn = (typeName) => {
  return (instance) => {
    return addRelayTypePromoterToInstance(typeName, instance);
  };
};

export const addRelayTypePromoterToListFn = (typeName) => {
  return (list) => {
    return addRelayTypePromoterToList(typeName, list);
  };
};

export const translateInstance = (
  entity,
  instance,
  { i18nLanguage, i18nDefaultLanguage, i18nLanguageDisableFallback },
) => {
  if (
    !instance ||
    !i18nLanguage ||
    i18nLanguage === i18nDefaultLanguage ||
    !entity.getI18nAttributeNames ||
    !entity.getI18nAttributeNames()
  ) {
    return instance;
  }

  const attributes = entity.getAttributes();
  const i18nAttributeNames = entity.getI18nAttributeNames();

  i18nAttributeNames.map((attributeName) => {
    const { required } = attributes[attributeName];
    const {
      fieldName: gqlFieldName,
      fieldNameI18n: gqlFieldNameI18n,
    } = getRegisteredEntityAttribute(
      entity.name,
      attributes[attributeName].name,
    );

    const translation = instance[gqlFieldNameI18n]
      ? instance[gqlFieldNameI18n][i18nLanguage]
      : undefined;

    if (i18nLanguageDisableFallback) {
      instance[gqlFieldName] =
        !isDefined(translation) && required
          ? new CustomError(
              `Translation for '${gqlFieldName}' not found`,
              'TranslationNotFoundError',
            )
          : translation;
    } else if (isDefined(translation)) {
      instance[gqlFieldName] = translation;
    }
  });

  return instance;
};

export const translateList = (entity, list, context) => {
  return list.map((instance) => {
    return translateInstance(entity, instance, context);
  });
};

export const translateInstanceFn = (entity, context) => {
  return (instance) => {
    return translateInstance(entity, instance, context);
  };
};

export const translateListFn = (entity, context) => {
  return (list) => {
    return translateList(entity, list, context);
  };
};

// emulate gql tag for prettier
export const gql = String.raw;
