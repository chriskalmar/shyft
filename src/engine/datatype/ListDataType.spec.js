import ListDataType, { isListDataType } from './ListDataType';

import { passOrThrow } from '../util';

import { DataTypeID, DataTypeString } from './dataTypes';

import { Language } from '../models/Language';

describe('ListDataType', () => {
  it('should have a name', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new ListDataType();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have a description', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new ListDataType({
        name: 'example',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should have an item type', () => {
    function fn() {
      // eslint-disable-next-line no-new
      new ListDataType({
        name: 'Example',
        description: 'Just some description',
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it("should return it's name", () => {
    const list = new ListDataType({
      name: 'someListDataTypeName',
      description: 'Just some description',
      itemType: DataTypeString,
    });

    expect(list.name).toBe('someListDataTypeName');
    expect(String(list)).toBe('someListDataTypeName');
  });

  it('should accept only valid item types', () => {
    // eslint-disable-next-line no-new
    new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: DataTypeString,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: Language,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: new ListDataType({
        name: 'Another',
        description: 'Just some description',
        itemType: DataTypeString,
      }),
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: () => DataTypeString,
    });

    function fn() {
      // eslint-disable-next-line no-new
      new ListDataType({
        name: 'Example',
        description: 'Just some description',
        itemType: [ 2, 7, 13 ],
      });
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should accept only valid item lenght ranges', () => {
    const definition = {
      name: 'Example',
      description: 'Just some description',
      itemType: DataTypeString,
    };

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 0,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 1,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 0,
      maxItems: 0,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 0,
      maxItems: 1,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 3,
      maxItems: 100,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 10,
      maxItems: 10,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      minItems: 10,
      maxItems: 0,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      maxItems: 0,
    });

    // eslint-disable-next-line no-new
    new ListDataType({
      ...definition,
      maxItems: 1,
    });

    function fn1() {
      // eslint-disable-next-line no-new
      new ListDataType({
        ...definition,
        minItems: -1,
      });
    }

    function fn2() {
      // eslint-disable-next-line no-new
      new ListDataType({
        ...definition,
        maxItems: -1,
      });
    }

    function fn3() {
      // eslint-disable-next-line no-new
      new ListDataType({
        ...definition,
        minItems: 2,
        maxItems: 1,
      });
    }

    expect(fn1).toThrowErrorMatchingSnapshot();
    expect(fn2).toThrowErrorMatchingSnapshot();
    expect(fn3).toThrowErrorMatchingSnapshot();
  });

  it('should accept valid dynamic item types', () => {
    // eslint-disable-next-line no-new
    const list1 = new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: () => DataTypeString,
    });

    // eslint-disable-next-line no-new
    const list2 = new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: () => Language,
    });

    // eslint-disable-next-line no-new
    const list3 = new ListDataType({
      name: 'Example',
      description: 'Just some description',
      itemType: () => {
        return new ListDataType({
          name: 'Another',
          description: 'Just some description',
          itemType: DataTypeString,
        });
      },
    });

    list1.getItemType();
    list2.getItemType();
    list3.getItemType();
  });

  it('should reject invalid dynamic item types', () => {
    function fn() {
      const list = new ListDataType({
        name: 'Example',
        description: 'Just some description',
        itemType: () => 'test',
      });

      list.getItemType();
    }

    expect(fn).toThrowErrorMatchingSnapshot();
  });

  it('should cache the dynamic item type after initial processing', () => {
    const list1 = new ListDataType({
      name: 'SomeName',
      description: 'Just some description',
      itemType: () => DataTypeString,
    });

    const itemType1 = list1.getItemType();
    const itemType1Again = list1.getItemType();

    expect(itemType1).toEqual(itemType1Again);

    const list2 = new ListDataType({
      name: 'SomeName',
      description: 'Just some description',
      itemType: DataTypeString,
    });

    const itemType2 = list2.getItemType();
    const itemType2Again = list2.getItemType();

    expect(itemType2).toEqual(itemType2Again);
  });

  describe('isListDataType', () => {
    it('should recognize objects of type ListDataType', () => {
      const list = new ListDataType({
        name: 'someListDataTypeName',
        description: 'Just some description',
        itemType: DataTypeID,
      });

      function fn() {
        passOrThrow(isListDataType(list), () => 'This error will never happen');
      }

      expect(fn).not.toThrow();
    });

    it('should recognize non-ListDataType objects', () => {
      function fn() {
        passOrThrow(
          isListDataType({}) ||
            isListDataType(function test() {}) ||
            isListDataType(Error),
          () => 'Not a ListDataType object',
        );
      }

      expect(fn).toThrowErrorMatchingSnapshot();
    });
  });
});
