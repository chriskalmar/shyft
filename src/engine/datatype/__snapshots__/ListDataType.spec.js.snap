// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListDataType isListDataType should recognize non-ListDataType objects 1`] = `"Not a ListDataType object"`;

exports[`ListDataType should accept only valid item lenght ranges 1`] = `"List data type 'Example' has invalid minItems setting '-1'"`;

exports[`ListDataType should accept only valid item lenght ranges 2`] = `"List data type 'Example' has invalid maxItems setting '-1'"`;

exports[`ListDataType should accept only valid item lenght ranges 3`] = `"List data type 'Example' has a bigger minItems than the maxItems setting"`;

exports[`ListDataType should accept only valid item types 1`] = `"List data type 'Example' has invalid item type '2,7,13'"`;

exports[`ListDataType should have a description 1`] = `"Missing description for list data type 'example'"`;

exports[`ListDataType should have a name 1`] = `"Missing list data type name"`;

exports[`ListDataType should have an item type 1`] = `"Missing item type for list data type 'Example'"`;

exports[`ListDataType should reject invalid dynamic item types 1`] = `"List data type 'Example' has invalid dynamic item type 'test'"`;
